### PR TITLE
[ASP.NET Core] Remove reflection for diagnostic source event payload

### LINKED ISSUE #3495: Instrumentation libraries improvements
# Instrumentation libraries improvements

## Subscribing to DiagnosticSource events

Instrumentation libraries subscribe to the diagnostic source events by looking at event names
that ends with Start, Stop, Exception etc [ref](https://github.com/open-telemetry/opentelemetry-dotnet/blob/6f2b1a0b3412f2860d065c81b058441977927584/src/OpenTelemetry/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs#L52-L67). This could result in subscribing to events which we did not
intend to. One such example is #3475.

**Discussion**: Should we match with explicit names rather than pattern matching?

## Relying on `Activity.Current`

When the diagnostic source events are written by the respective framework libraries. Usually the activity is 
not part of the payload or activity itself is not started by the framework e.g. SQL instrumentation. In order
to enrich/modify activities instrumentation libraries rely on `Activity.Current` [ref](https://github.com/open-telemetry/opentelemetry-dotnet/blob/6f2b1a0b3412f2860d065c81b058441977927584/src/OpenTelemetry/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs#L52-L67) which may cause issues e.g. #3490. 

**Action** : Review any more cases like the one shown in example.

## Documentation

We do not have a detailed inventory of events that the instrumentation libraries in this
repo subscribes to. Having this documented for each instrumentation would help provide more clarity to users. Also, this would help us analyze and make decision on whether we need to support more events or update any existing ones. 

**Action**: Update `Readme` documents for each instrumentation library with the event names it subscribes to.

## Avoid reflection (ASP.NET Core)

Reflection is used to extract properties from events payload ([e.g.](https://github.com/open-telemetry/opentelemetry-dotnet/blob/44ac86cb741544b4e5b5ff043bb71454b6f9d8ae/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L251-L255)). This was done in order to avoid nuget reference. Now that we have FrameworkReference `Microsoft.AspNetCore.App`, could we avoid reflection?

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Partially Fixes #3495 

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 1]
```diff
@@ -76,7 +72,7 @@ public override void OnStartActivity(Activity activity, object payload)
                 return;
             }
 
-            _ = this.startContextFetcher.TryFetch(payload, out HttpContext context);
+            HttpContext context = payload as DefaultHttpContext;
```
Comment (alanwest) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:75]: I don't think this is guaranteed to always be of type `DefaultHttpContext`. Might be a bit of an unusual case but I share this example here https://github.com/open-telemetry/opentelemetry-dotnet/pull/3498#discussion_r932771470. `DefaultHttpContext` is derived from `HttpContext`.

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:75]: Not sure if I understood. With the example (adding HttpContextFactory), I am getting the payload as DefaultHttpContext type.

Comment (alanwest) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:75]: Oh yea you're right, that example uses the DefaultHttpContextFactory. My presumption is that you could use a different factory that might generate a totally different type.

Comment (alanwest) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:75]: Not a fully working example, but something like 

```csharp
private class TestHttpContextFactory : IHttpContextFactory
 {
    public HttpContext Create(IFeatureCollection featureCollection)
    {
        return new MyHttpContext();
    }
}

private class MyHttpContext : HttpContext
{
}
```

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs]: changed to HttpContext.

**Review by alanwest — APPROVED** [review]
Like a 🧛 whose reflection can't be seen in a 🪞. I no longer see reflection in this PR.

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Removes reflection for extracting required objects from DiagnosticSource event
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md:5]: no user facing change here, so we dont need this changelog right?

**Review by cijothomas — APPROVED** [review]
LGTM. wondering if it is time to resurrect the perf benchmarks and see whether these have any impact on them.

File: src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md (hunk 2 above)
Comment (reyang) [reply src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md:5]: +1

**Pull request merged**